### DISCUSSION TITLE
iOS16 네비게이션 링크 이슈 해결방법..!

### DISCUSSION THREAD

**Discussion by JIWON1923** [Team-ShortcutsZip/MacC-Team-HappyAnding#389] [Polls]
### 의견이 필요합니다.

#380 참고해주세요!

해당 이슈가 iOS 16에서 발생하는 이슈인 것이 확인되었습니다!
[애플 프로젝트](https://developer.apple.com/documentation/swiftui/bringing_robust_navigation_structure_to_your_swiftui_app)에서도 동일한 이슈가 발생합니다.
iOS 16.1부터는 해당 이슈가 발생하지 않습니다.

### 대안

1. available 이용해서 Navigation Stack과 NavigationView로 해서 관리하는 방법

2. iOS16에서 half swipe back 제스처를 커스텀 하는 방법
  - 이 방법은 조금이라도 swipe하면 뷰를 pop 시키는 방법입니다! (제일 가능성이 높아보여서 구현해봤는데 문제가 심해유)
  - 단점은 코드가 복잡해질거예요 (available을 여러 곳에서 사용해야합니다.)
  - 제일 문제는 제스처가 굉장히 부자연스러워요!! (하단 영상에 문제상황 첨부합니다)

3. 최소 버전을 iOS 16.1로 수정하는 방법
  - 최악의 최악의 경우로만 사용해야합니다.
  
추가 해결방법 알고 계시면 공유 부탁드립니다!
1번 방법을 사용한다면, available로 관리하게 되는데, 중복 코드를 줄이기 위해 뷰빌더를 사용해야할 것 같습니다! (꽤 많은 수정이 일어날 것 같아요)
2번 방법을 구현은 해 봤는데, 밑에 영상과 설명 참고해주세요!

### 제스처 커스텀 문제상황 공유
- Back button을 hidden 시켜서 swipe 제스처를 막아야합니다.
  - `NavigationBarColorModifier`를 available(iOS 16.1, *)로 설정했습니다.
  - ListShortcutsView와 ReadShortcutsVIew의 back button을 hidden 시켰습니다.

- 동영상 참고 사항
  - backbutton, navigation 컬러 등은 고려하지 말아주세요!
  - 두 화면 모두 같은 제스처를 한 결과입니다!
  - 뷰 자체가 pop 되는 애니메이션은 자연스럽지만, 손가락을 모두 뗀 후에 뷰가 Pop되는 모습이 부자연스럽다는 제 의견입니다! 여러분은 어떠신가유

- 문제사항
  - 카테고리에서 swipe back이 동작하지 않습니다. 
  - 드래그 제스처를 통해 뷰를 pop시키는데, 스와이프를 한 후 손가락을 화면에서 떼면 버튼 클릭으로 이어집니다... 
  - 그래서 readShortcutsView에서는 pop이 잘 되지만, 카테고리에서는 pop이 제대로 되지 않습니다 (투표 결과 보고 방법을 찾아보긴 할게요.,.)
  - 버튼이 없는 위치에서 제스처를 수행하면 Pop이 됩니다 (영상에는 첨부되지 않았어요)

https://user-images.githubusercontent.com/68676844/217001950-67e1f818-c4ce-4131-b7ca-0e75a185d8e0.mov

**Comment by JMM00** [comment]
저는 화면이 pop되는 상태보다 더 문제라고 느끼는 부분이 사용자의 입장에서 안드로이드처럼 하단 버튼도 없는 상태에서 뒤로가기 버튼이 비활성화되면 더이상 뒤로 갈 수 없는 화면으로 보여져서 사용하는데 불편함이 있을 것 같아 2번은 제외하는 것이 좋을 것 같습니다.

다만 앱이 제대로 동작하지 않는 문제기 때문에 navigation stack과 navigation view로 분리하여 해결이 가능하다면 이 방향으로라도 해결한 후 더 좋은 방법을 찾아보는 것은 어떨까요..?

**Comment by JIWON1923** [comment]
> 저는 화면이 pop되는 상태보다 더 문제라고 느끼는 부분이 사용자의 입장에서 안드로이드처럼 하단 버튼도 없는 상태에서 뒤로가기 버튼이 비활성화되면 더이상 뒤로 갈 수 없는 화면으로 보여져서 사용하는데 불편함이 있을 것 같아 2번은 제외하는 것이 좋을 것 같습니다.
> 
> 
> 
> 다만 앱이 제대로 동작하지 않는 문제기 때문에 navigation stack과 navigation view로 분리하여 해결이 가능하다면 이 방향으로라도 해결한 후 더 좋은 방법을 찾아보는 것은 어떨까요..?


 음... 백버튼이 없어서 혼란을 느낄 것 같다는 이야기가 맞을까요? (제스처가 작동하지 않아서 스택과 뷰를 구분하자는 말씀이라면 밑에 의견은 쓰루 하셔도 됩니다!)

 뒤로가기 버튼이 비활성화 되는 것은 아니에요.
현재 백버튼을 히든 시킨 이유는 swipe back 제스처를 비활성화하기 위해서 넣어둔 코드입니다. (만약 2번 방식으로 구현한다면, 추후 백버튼 삽입 코드가 들어갈 예정이에요. 그래서 부자연스러운 제스처를 공유하기 위한 프로토타입이라고 생각해주시면 됩니다.)

 정리하자면 2번 방식으로 구현한다면, iOS16에서는 기본 백버튼이 아닌 커스텀 백버튼을 제공하는 코드가 들어갈 예정이에요 (드래그 제스처를 커스텀 하는 방식이죠!)

 그래서 영상 보실 때 제스처 위주로 확인해주시면 될 것 같아요!! (물론 저도 분리가 제일 나을 것 같음^^;;)

**Reply by JMM00** [reply]
백버튼은 고려하지 말라는 말이 이런 의미였군요.. 이해했습니다.
ReadShortcutView에서 주로 제스처를 실행하게 되는 하단 부분이 내부 탭을 변경하는 스와이프 제스쳐와 중복되기 때문에 (특히 댓글 부분에 있는 경우) 화면이 손에 맞춰서 바로 따라오지 않는 것이 약간의 혼동을 줄 수 있을 것 같습니다. 이 부분이 보완된다면 저는 애니메이션 자체는 크게 어색하다고 느껴지지는 않습니다.

**Reply by JIWON1923** [reply]
화면이 손에 맞춰서 바로 따라오는 제스처 = "손가락을 모두 뗀 후에 뷰가 Pop되는 모습"라고 보시면 될 것 같아요.

화면이 손에 맞춰서 바로 따라오기 때문에 pop하는 행위를 취소할 수 있게 되는거고, 해당 작업이 iOS 16 네비게이션 스택에서 문제가 되는 것입니다!
그래서 2번 같은 경우는 그 제스처를 제거해서 pop을 취소할 수 있는 가능성을 배제하는 방식이에요.

결론적으로 해당 부분은 보완할 수 없을 것 같습니다.(또는 다른 해결방법 생각나시면 공유부탁드려요!) 그래서 현재 화면에서 UI말고 모든 UX만 봐주시면 됩니다. (제스처와 애니메이션의 연결부분)

지민님의 의견이 손의 움직임에 따라 뷰가 따라오지 않는다면, 2번 방식은 제외하자는 의견이 맞을까요?

**Reply by JMM00** [reply]
넵 제스쳐 중복으로 제스쳐가 제대로 실행되지 않는 경우가 간헐적으로 존재해서 보완되지 않는다면 2번은 제외하는 것이 좋을 것 같습니다.

**Comment by HanGyeongjun** [comment]
2번처럼 제스처를 제거하는 것 보다는 1번이 UX 측면에서 더 적합할 것 같습니다.
그런데 만약 16.0에서 NavigationView를 사용하는 방향으로 가져간다면 광범위한 코드 수정이 필요할 것 같은데, 1번과 2번 중 어떤 방법이 공수가 더 많이 들까요? 2번도 크게 UX를 해친다고는 생각하지 않아서 둘 중 공수가 덜 드는 방향으로 가져가는 것도 괜찮다고 생각합니다!

**Reply by HanGyeongjun** [reply]
둘 다 많이 든다면 그냥 놔두는것도...😶🤫

**Reply by JIWON1923** [reply]
공수는 둘 다 비슷할것 같아요
다만 1번의 경우 유지보수를 고려하며 스택과 뷰를 잘 분리할 수 있을지 고민하는 시간이 길어질 것 같습니다😂

**Reply by HanGyeongjun** [reply]
글쿤요... 다른 분들 의견 더 들어보고 종합적으로 결정해야 할 것 같네요
